### ISSUE TITLE
Can't work in SBT tasks: java.lang.NoClassDefFoundError: scala/tools/scalap/scalax/rules/scalasig/ScalaSigSymbol

### ISSUE THREAD

**Issue by freewind** [json4s/json4s#236]
``` scala
java.lang.NoClassDefFoundError: scala/tools/scalap/scalax/rules/scalasig/ScalaSigSymbol
```

http://stackoverflow.com/questions/30229861/when-i-convert-an-object-to-json-with-json4s-in-a-sbt-task-it-reports-errors-h

And similar:

https://github.com/orrsella/sbt-sublime/issues/11

**Comment by seratch** [comment]
json4s has its own scalap fork since 3.3.0. 

https://github.com/json4s/json4s/tree/3.3/scalap/

If you have some troubles related with scalap in 3.2.x, please make a pull request toward the `3.2` branch.

**Comment by freewind** [comment]
You can see I've add that:

https://github.com/freewind/json4s-210-test/blob/master/project/plugins.sbt#L4

Even with `scalap`, it has the same error.

**Comment by seratch** [comment]
I've not figured out why it is yet, but your sbt project depends on json4s 3.2.x without a doubt.

- https://github.com/json4s/json4s/blob/v3.3.0.RC2/core/src/main/scala/org/json4s/reflect/Reflector.scala#L123
- https://github.com/json4s/json4s/blob/v3.2.11/core/src/main/scala/org/json4s/reflect/Reflector.scala#L123

```
java.lang.NoClassDefFoundError: scala/tools/scalap/scalax/rules/scalasig/ScalaSigSymbol
    at org.json4s.reflect.Reflector$ClassDescriptorBuilder$$anonfun$constructorsAndCompanion$3$$anonfun$13.apply(Reflector.scala:123)
    at org.json4s.reflect.Reflector$ClassDescriptorBuilder$$anonfun$constructorsAndCompanion$3$$anonfun$13.apply(Reflector.scala:122)
```

**Comment by xuwei-k** [comment]
You can't use sbt 0.13.8 and json4s 3.3.0.RC2 together, because sbt 0.13.8 depends on json4s 3.2.10.

- https://github.com/sbt/sbt/blob/v0.13.8/project/Dependencies.scala#L15
- https://github.com/sbt/serialization/blob/v0.1.1/project/Dependencies.scala#L16
- https://twitter.com/eed3si9n/status/535140774145687552

**Comment by eed3si9n** [comment]
oops. Json4s *is* used by sbt as well as sbt/serialization project. AFAIK, Json4s's classes are not exposed via our API, so sbt might be able to fork/shade the AST in a future release. /cc @jsuereth

**Comment by jsuereth** [comment]
I guess teh question is, is Json4s relatively binary compatible or not?  If not, we may need to shade/pull it out.

**Comment by rossabaker** [comment]
Mima is being added for 3.3, but there was churn in patch releases before
that point.

There is also a new proposal for a 4.x, which would be another bincompat
event if it comes to fruition.

**Comment by eed3si9n** [comment]
We use Jawn to parse, so only thing I am using out of json4s I think is pure AST. I was hoping json4s to be the common AST project, but if it's changing frequently we'd have to fork and copy-pasta. Would you guys be open to creating an AST only JAR that doesn't change for 2 years, kind of thing, or is it not your project goal?

**Comment by eed3si9n** [comment]
If the shaded `org.json4s.v3.JValue` is in your JAR and and type alias it to `org.json4s.JValue`,
you can add a new `org.json4s.v4.JValue` while keeping the old one around forever. Not sure if that'll completely isolate future breakage by eviction.

**Comment by jsuereth** [comment]
So, I think we could probably just shade in sbt-serialization.  HOwever, I would love to have a pure API jar that just never changes.  We're not really using much outside the pure API, we even provide our own implementation of JsonMethods so we could avoid BC breakage...

**Comment by rossabaker** [comment]
I speak as a very minor contributor on this project, but one who has been around since the beginning.

There is a [json4s-ast](https://github.com/json4s/json4s/tree/v3.3.0.RC3/ast/src/main/scala/org/json4s) module, on which json4s-core depends.  That was extracted around the time some Typesafe folks and json4s folks had talks about a renewed effort to make a de facto standard JSON AST in Scala.  We still have play-json, spray-json, json4s-ast, and that's just the libraries from the people who were in unification talks.

I would also love to see a lean, stable AST.  I don't see good reason that `Diff`, `Merge`, or `DSL` belong in json-ast.  I vaguely recall this was an effort to maintain compatibility with code dating back to our lift-json origins.  Perhaps we could agree on and publish a lean AST, get jawn parsing to it, and then codecs and typeclass instances and all the other trappings of a richer JSON experience could follow and evolve at their own pace.

**Comment by eed3si9n** [comment]
> There is a json4s-ast module, on which json4s-core depends. 

ok. So I share some of the responsibility because I introduced dependency to json4s-core.
https://github.com/sbt/serialization/blob/843e5e805340be2f5fd93c7db3f9af09c19d1e05/project/Dependencies.scala#L16

Now that we have our own JSON writer, I should try to remove the dependency to json4s-core.

**Comment by jeffreyolchovy** [comment]
What is the status on this? Is it still not possible to use json4s in sbt tasks? I am hitting the same issue trying any recent version of sbt combined with any version of json4s.

**Comment by imarios** [comment]
Hitting the same here as well. Have an sbt plugin with dependency to json4s and gives me the same issue. Any suggestion?

**Comment by jeffreyolchovy** [comment]
Personally, to work around this problem I started publishing an identical artifact with a modified package name (i.e. something like `org.json4s` to `org.json4me`). As a first course of action, I simply tried to [jarjar](https://code.google.com/archive/p/jarjar/) the `json4s` artifacts, but that didn't work, so I did the transform manually. Not ideal in the least, but it works...

**Comment by imarios** [comment]
thank you @jeffreyolchovy ! I am trying to workout this using shading in sbt assembly and publishing my stuff with a shaded json4s. I think this is very similar to what you suggest. thanks!

**Comment by imarios** [comment]
This seems to be more of am sbt issue. Hi Eugene, do you think there is a chance for this be addressed in sbt 1.0? @eed3si9n

**Comment by eed3si9n** [comment]
Yes. We're in the process of switching sbt 1.0 to SLIP-28 Scala JSON.

**Comment by magnolia-k** [comment]
This issue will be closed as sbt 1.x does not depend on Json4s.
If there are any other issues, please raise another issue or reopen this issue.

https://github.com/sbt/sbt/blob/c3f016829e5daccce77c5473c7a9b1757e385053/project/Dependencies.scala#L119
